### ISSUE TITLE
"Root resource was present, now absent" errors during apply

### ISSUE THREAD

**Issue by annettejanewilson** [newrelic/terraform-provider-newrelic#1865]
## Overview

We've seen this happen a couple of times spontaneously. We receive an error like the following:

```
Error: Provider produced inconsistent result after apply

When applying changes to
newrelic_synthetics_multilocation_alert_condition.cdn-month-view-page,
provider "provider[\"registry.terraform.io/newrelic/newrelic\"]" produced an
unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
```

## Interpretation

It appears that the plan calls for a resource to be created, and during the apply an attempt is made to create it, but we don't seem to see any message to say what happened to the attempt, we just get this error that the new state is unexpected because the resource that should have been created doesn't exist. In both cases where we observed this, a subsequent plan and apply resolved the issue.

## Logs

Here's the output of the plan operation:

```
Initializing modules...
- client_errors_static_alert in ../../../src/modules/web_enablement/client_errors_alert

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.

Initializing provider plugins...
- Finding newrelic/newrelic versions matching "~> 2.27"...
- Installing newrelic/newrelic v2.48.0...
- Installed newrelic/newrelic v2.48.0 (signed by a HashiCorp partner, key ID DC9FC6B1FCE47986)

Partner and community providers are signed by their developers.
If you'd like to know more about provider signing, you can read about it here:
https://www.terraform.io/docs/cli/plugins/signing.html

Terraform has created a lock file .terraform.lock.hcl to record the provider
selections it made above. Include this file in your version control repository
so that Terraform can guarantee to make the same selections by default when
you run "terraform init" in the future.

Terraform has been successfully initialized!

You may now begin working with Terraform. Try running "terraform plan" to see
any changes that are required for your infrastructure. All Terraform commands
should now work.

If you ever set or change modules or backend configuration for Terraform,
rerun this command to reinitialize your working directory. If you forget, other
commands will detect it and remind you to do so if necessary.

newrelic_alert_policy.akamai-month-view-page: Refreshing state... [id=2338913]
newrelic_alert_policy.client_uncaught_errors_baseline: Refreshing state... [id=2858574]
newrelic_alert_policy_channel.akamai-month-view-page: Refreshing state... [id=2338913:6530256]
newrelic_alert_muting_rule.mute-client_errors_baseline: Refreshing state... [id=3117610:16367300]
newrelic_synthetics_monitor.akamai-month-view-page: Refreshing state... [id=8790ce00-c4ab-4863-ae0a-4a84ce7000e9]
newrelic_synthetics_multilocation_alert_condition.akamai-month-view-page: Refreshing state... [id=2338913:24736738]
newrelic_alert_muting_rule.mute-low_hotels_cta_clicks_baseline: Refreshing state... [id=3117610:17485467]
newrelic_alert_muting_rule.mute-low_flights_cta_clicks_static: Refreshing state... [id=3117610:16368599]
newrelic_alert_channel.tmp-slack-month-view-alerts: Refreshing state... [id=6416892]
newrelic_alert_policy.client_errors_static: Refreshing state... [id=2858575]
newrelic_alert_policy.client_errors_baseline: Refreshing state... [id=2858579]
newrelic_alert_policy.low_hotels_cta_clicks_baseline: Refreshing state... [id=3103687]
newrelic_alert_policy.low_flights_cta_clicks_static: Refreshing state... [id=2858850]
newrelic_alert_channel.slack-month-view-alerts: Refreshing state... [id=6416893]
newrelic_alert_policy.client_uncaught_errors_static: Refreshing state... [id=2858577]
newrelic_alert_policy.low_flights_cta_clicks_baseline: Refreshing state... [id=3103688]
newrelic_alert_muting_rule.mute-low_flights_cta_clicks_baseline: Refreshing state... [id=3117610:17485465]
newrelic_alert_muting_rule.mute-client_uncaught_errors_baseline: Refreshing state... [id=3117610:16367299]
newrelic_alert_policy.low_hotels_cta_clicks_static: Refreshing state... [id=2858849]
newrelic_alert_muting_rule.mute-low_hotels_cta_clicks_static: Refreshing state... [id=3117610:16368598]
newrelic_alert_muting_rule.mute-client_uncaught_errors_static: Refreshing state... [id=3117610:16367301]
newrelic_alert_channel.vo-month-view-critical: Refreshing state... [id=6530256]
newrelic_alert_muting_rule.mute-client_errors_static: Refreshing state... [id=3117610:16367302]
newrelic_nrql_alert_condition.low_flights_cta_clicks_static: Refreshing state... [id=2858850:25688037]
newrelic_nrql_alert_condition.client_uncaught_errors_baseline: Refreshing state... [id=2858574:25687619]
newrelic_nrql_alert_condition.client_uncaught_errors_static: Refreshing state... [id=2858577:25687620]
newrelic_nrql_alert_condition.client_errors_baseline: Refreshing state... [id=2858579:25687622]
newrelic_nrql_alert_condition.low_flights_cta_clicks_baseline: Refreshing state... [id=3103688:26149627]
module.client_errors_static_alert.newrelic_nrql_alert_condition.client_errors_alert: Refreshing state... [id=2858575:25687621]
newrelic_nrql_alert_condition.low_hotels_cta_clicks_baseline: Refreshing state... [id=3103687:26149628]
newrelic_alert_policy_channel.low_hotels_cta_clicks_baseline: Refreshing state... [id=3103687:6416893]
newrelic_alert_policy_channel.low_flights_cta_clicks_baseline: Refreshing state... [id=3103688:6416893]
newrelic_alert_policy_channel.client_uncaught_errors_static: Refreshing state... [id=2858577:6416893]
newrelic_alert_policy_channel.client_uncaught_errors_baseline: Refreshing state... [id=2858574:6416893]
newrelic_alert_policy_channel.client_errors_static: Refreshing state... [id=2858575:6416893]
newrelic_alert_policy_channel.client_errors_baseline: Refreshing state... [id=2858579:6416893]
newrelic_alert_policy_channel.low_flights_cta_clicks_static: Refreshing state... [id=2858850:6416893:6530256]
newrelic_nrql_alert_condition.low_hotels_cta_clicks_static: Refreshing state... [id=2858849:25688038]
newrelic_alert_policy_channel.low_hotels_cta_clicks_static: Refreshing state... [id=2858849:6416893:6530256]

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create
  - destroy

Terraform will perform the following actions:

  # newrelic_alert_policy.akamai-month-view-page will be destroyed
  # (because newrelic_alert_policy.akamai-month-view-page is not in configuration)
  - resource "newrelic_alert_policy" "akamai-month-view-page" {
      - account_id          = 3117610 -> null
      - id                  = "2338913" -> null
      - incident_preference = "PER_POLICY" -> null
      - name                = "month-view-page:akamai-month-view-page" -> null
    }

  # newrelic_alert_policy.cdn-month-view-page will be created
  + resource "newrelic_alert_policy" "cdn-month-view-page" {
      + account_id          = (known after apply)
      + id                  = (known after apply)
      + incident_preference = "PER_POLICY"
      + name                = "month-view-page:cdn-month-view-page"
    }

  # newrelic_alert_policy_channel.akamai-month-view-page will be destroyed
  # (because newrelic_alert_policy_channel.akamai-month-view-page is not in configuration)
  - resource "newrelic_alert_policy_channel" "akamai-month-view-page" {
      - channel_ids = [
          - 6530256,
        ] -> null
      - id          = "2338913:6530256" -> null
      - policy_id   = 2338913 -> null
    }

  # newrelic_alert_policy_channel.cdn-month-view-page will be created
  + resource "newrelic_alert_policy_channel" "cdn-month-view-page" {
      + account_id  = (known after apply)
      + channel_ids = [
          + 6530256,
        ]
      + id          = (known after apply)
      + policy_id   = (known after apply)
    }

  # newrelic_synthetics_monitor.akamai-month-view-page will be destroyed
  # (because newrelic_synthetics_monitor.akamai-month-view-page is not in configuration)
  - resource "newrelic_synthetics_monitor" "akamai-month-view-page" {
      - bypass_head_request       = true -> null
      - frequency                 = 1 -> null
      - id                        = "8790ce00-c4ab-4863-ae0a-4a84ce7000e9" -> null
      - locations                 = [
          - "AWS_EU_CENTRAL_1",
          - "AWS_EU_WEST_2",
          - "AWS_US_WEST_1",
        ] -> null
      - name                      = "month-view-page:akamai-month-view-page" -> null
      - sla_threshold             = 7 -> null
      - status                    = "ENABLED" -> null
      - treat_redirect_as_failure = false -> null
      - type                      = "SIMPLE" -> null
      - uri                       = "https://www.skyscanner.net/transport/flights/edi/lond/" -> null
      - validation_string         = "month-view-page" -> null
      - verify_ssl                = false -> null
    }

  # newrelic_synthetics_monitor.cdn-month-view-page will be created
  + resource "newrelic_synthetics_monitor" "cdn-month-view-page" {
      + bypass_head_request = true
      + frequency           = 1
      + id                  = (known after apply)
      + locations           = [
          + "AWS_EU_CENTRAL_1",
          + "AWS_EU_WEST_2",
          + "AWS_US_WEST_1",
        ]
      + name                = "month-view-page:cdn-month-view-page"
      + sla_threshold       = 7
      + status              = "ENABLED"
      + type                = "SIMPLE"
      + uri                 = "https://www.skyscanner.net/transport/flights/edi/lond/"
      + validation_string   = "month-view-page"
    }

  # newrelic_synthetics_multilocation_alert_condition.akamai-month-view-page will be destroyed
  # (because newrelic_synthetics_multilocation_alert_condition.akamai-month-view-page is not in configuration)
  - resource "newrelic_synthetics_multilocation_alert_condition" "akamai-month-view-page" {
      - enabled                      = true -> null
      - entities                     = [
          - "8790ce00-c4ab-4863-ae0a-4a84ce7000e9",
        ] -> null
      - id                           = "2338913:24736738" -> null
      - name                         = "month-view-page:akamai-month-view-page" -> null
      - policy_id                    = 2338913 -> null
      - violation_time_limit_seconds = 86400 -> null

      - critical {
          - threshold = 2 -> null
        }
    }

  # newrelic_synthetics_multilocation_alert_condition.cdn-month-view-page will be created
  + resource "newrelic_synthetics_multilocation_alert_condition" "cdn-month-view-page" {
      + enabled                      = true
      + entities                     = (known after apply)
      + id                           = (known after apply)
      + name                         = "month-view-page:cdn-month-view-page"
      + policy_id                    = (known after apply)
      + violation_time_limit_seconds = 86400

      + critical {
          + threshold = 2
        }
    }

Plan: 4 to add, 0 to change, 4 to destroy.

─────────────────────────────────────────────────────────────────────────────

Saved the plan to: /atlantis-data/repos/newrelic/newrelic-tf/573/default/config/production/month-view-page/default.tfplan

To perform exactly these actions, run the following command to apply:
    terraform apply "/atlantis-data/repos/newrelic/newrelic-tf/573/default/config/production/month-view-page/default.tfplan" 
```

Here's the apply:

```
exit status 1: running "terragrunt apply -no-color $PLANFILE" in "/atlantis-data/repos/newrelic/newrelic-tf/573/default/config/production/month-view-page": 
newrelic_alert_policy_channel.akamai-month-view-page: Destroying... [id=2338913:6530256]
newrelic_synthetics_multilocation_alert_condition.akamai-month-view-page: Destroying... [id=2338913:24736738]
newrelic_alert_policy.cdn-month-view-page: Creating...
newrelic_synthetics_monitor.cdn-month-view-page: Creating...
newrelic_synthetics_multilocation_alert_condition.akamai-month-view-page: Destruction complete after 1s
newrelic_synthetics_monitor.akamai-month-view-page: Destroying... [id=8790ce00-c4ab-4863-ae0a-4a84ce7000e9]
newrelic_alert_policy_channel.akamai-month-view-page: Destruction complete after 1s
newrelic_alert_policy.akamai-month-view-page: Destroying... [id=2338913]
newrelic_alert_policy.cdn-month-view-page: Creation complete after 1s [id=3251959]
newrelic_alert_policy_channel.cdn-month-view-page: Creating...
newrelic_alert_policy_channel.cdn-month-view-page: Creation complete after 1s [id=3251959:6530256]
newrelic_alert_policy.akamai-month-view-page: Destruction complete after 1s
newrelic_synthetics_monitor.akamai-month-view-page: Destruction complete after 2s
newrelic_synthetics_monitor.cdn-month-view-page: Creation complete after 3s [id=8f13e728-fd46-4625-82d6-025a76d3f4f2]
newrelic_synthetics_multilocation_alert_condition.cdn-month-view-page: Creating...
newrelic_synthetics_multilocation_alert_condition.cdn-month-view-page: Still creating... [10s elapsed]

Error: Provider produced inconsistent result after apply

When applying changes to
newrelic_synthetics_multilocation_alert_condition.cdn-month-view-page,
provider "provider[\"registry.terraform.io/newrelic/newrelic\"]" produced an
unexpected new value: Root resource was present, but now absent.

This is a bug in the provider, which should be reported in the provider's own
issue tracker.
time=2022-06-22T07:21:22Z level=error msg=1 error occurred:
	* exit status 1



Pull Request approved by CODEOWNERS 
```

## Behaviour

The message seems to indicate that the plan produced by Terraform reported that there would be a resource `newrelic_synthetics_multilocation_alert_condition.cdn-month-view-page` after completion. However when Terraform executed the command, afterwards it looked and could not see that resource.

The outer message "provider X produced an unexpected new value: Y This is a bug in the provider..." comes from [Terraform's node_resource_abstract_instance.go][terraform-nrai].

The inner message "Root resource was present, but now absent" comes from [Terraform's compatible.go][terraform-compat].

Given that no "Creation complete after X" message was displayed I am speculating that the operation timed out or otherwise silently failed.

## Configuration

Here's the relevant configuration, `config/production/month-view-page/synthetics-cdn-month-view-page.tf`:

```hcl
resource "newrelic_synthetics_monitor" "cdn-month-view-page" {
  name      = "month-view-page:cdn-month-view-page"
  type      = "SIMPLE"
  frequency = 1
  status    = "ENABLED"
  locations = ["AWS_EU_WEST_2", "AWS_EU_CENTRAL_1", "AWS_US_WEST_1"]

  uri                 = "https://www.skyscanner.net/transport/flights/edi/lond/"
  validation_string   = "month-view-page"
  bypass_head_request = true
}

resource "newrelic_alert_policy" "cdn-month-view-page" {
  name = "month-view-page:cdn-month-view-page"
}

resource "newrelic_synthetics_multilocation_alert_condition" "cdn-month-view-page" {
  policy_id                    = newrelic_alert_policy.cdn-month-view-page.id
  name                         = "month-view-page:cdn-month-view-page"
  violation_time_limit_seconds = "86400"
  entities                     = [newrelic_synthetics_monitor.cdn-month-view-page.id]
  critical {
    threshold = 2
  }
}

resource "newrelic_alert_policy_channel" "cdn-month-view-page" {
  policy_id = newrelic_alert_policy.cdn-month-view-page.id
  channel_ids = [
    newrelic_alert_channel.vo-month-view-critical.id,
  ]
}
```

Here's `config/production/terragrunt.hcl`:
```hcl

   
generate "provider" {
  path      = "provider.tf"
  if_exists = "skip"
  contents = <<EOF
terraform {
  required_version = "~> 1.1.4"
  required_providers {
    newrelic = {
      source  = "newrelic/newrelic"
      version = "~> 2.27"
    }
  }
  backend "s3" {
    region         = "eu-west-1"
    bucket         = "skyscanner-prod-newrelic-tf-state"
    key            = "config/production/${path_relative_to_include()}/terraform.tfstate"
    dynamodb_table = "skyscanner-prod-newrelic-tf-state"
  }
}
locals {
  account_id = 3117610  # Skyscanner Production 
}
provider "newrelic" {
  account_id = local.account_id
  api_key = var.newrelic_api_key_prod
  region = "US"
}
EOF
}
 
generate "common_vars" {
  path      = "common_vars.tf"
  if_exists = "skip"
  contents = <<EOF
variable "newrelic_api_key_prod" {}
variable "vo_api_key" {}
variable "vo_nr_integration_key" {}
variable "vo_api_id" {}
variable "kong_api_key" {}
variable "slack_webhook" {}
EOF
}
```

Here's `config/production/month-page-view/terragrunt.hcl`:

```hcl
include {
  path = find_in_parent_folders()
}
```

I can provide the entire configuration but I don't think any of the rest of it is involved here, and I know that this doesn't trigger reliably so I am not expecting to be able to provide a reliable reproduction.

I believe we're running Atlantis 0.19.2 which uses Terraform 1.1.7. You can see from the output that this uses version 2.48.0 of the provider.

## Expected behaviour

Well, ideally it would succeed first time and not fail, or even spend a short time retrying. Failing that, a clear message indicating what operation failed and why would make this easier to interpret - the output just seems to jump from "I'm trying to create a thing..." to "Something is wrong, the thing does not exist!"

## Related

A number of [similar bugs][bug-search] seem to have been reported but generally closed as unreprodicible.

[terraform-nrai]: https://github.com/hashicorp/terraform/blob/2ad3ac7d016303928b368f40ea6e8cbf02a898e2/internal/terraform/node_resource_abstract_instance.go#L2187-L2231
[terraform-compat]: https://github.com/hashicorp/terraform/blob/c687ebeaf19c7c89188727ffc54b03bcc6e51a01/internal/plans/objchange/compatible.go#L33-L44
[bug-search]: https://github.com/newrelic/terraform-provider-newrelic/issues?q=is%3Aissue+present+absent

**Comment by kidk** [comment]
Thanks a lot @annettejanewilson for the very detailed bug report. 👍  We'll try to figure out what's going on here.

**Comment by kidk** [comment]
We're currently in the process of moving the Synthetics resources from our old REST api to new GraphQL API. As you've noticed from other tickets this is hard to debug from our side, so we want to wait until the migration has happened to see if the bug might disappear on its own. The REST API is known to have timeouts and race conditions that can lead to situations like this. 

A possible workaround might be to use NRQL alerts, but let me confirm that with our Synthetics team. 

Thanks for your patience.

**Comment by annettejanewilson** [comment]
Thank you, I appreciate your investigation.

**Comment by haydnchapman-pg** [comment]
Hi @kidk - has there been any headway on an actual solution for this issue, as we've just suddenly started seeing the same error message when running out our TF to create the channel policy (the actual module/resource name has been and replaced with `{resource}`)

```
Error: Provider produced inconsistent result after apply 

When applying changes to 
{resource},
provider "provider[\"registry.terraform.io/newrelic/newrelic\"]" produced an 
unexpected new value: Root resource was present, but now absent. 
This is a bug in the provider, which should be reported in the provider's own 
issue tracker. 
```

**Comment by darren-steven** [comment]
I'm seeing this behavior as well. It's a new behaviour - these simple TF plans (one policy, one alert) used to work well, but now fail

```
newrelic_alert_policy.tf_test_policy_nonprod: Creating...
newrelic_alert_policy.tf_test_policy_nonprod: Creation complete after 1s [id=3668521]
newrelic_nrql_alert_condition.foo_testing: Creating...
╷
│ Error: Provider produced inconsistent result after apply
│ 
│ When applying changes to newrelic_nrql_alert_condition.foo_testing, provider "provider[\"registry.terraform.io/newrelic/newrelic\"]" produced an unexpected new value: Root resource was present, but now absent.
│ 
│ This is a bug in the provider, which should be reported in the provider's own issue tracker.
╵
```
When I look at the results in NewRelic, I see the policy, but "most" times, it appears to be non-functional. I can eventually delete it, but NR is confused as to it's existence or not.

Terraform v1.3.3
on darwin_amd64
+ provider registry.terraform.io/hashicorp/time v0.9.0
+ provider registry.terraform.io/newrelic/newrelic v3.5.2

We se the same issue on gitlab as well

**Comment by yuklia** [comment]
Have same issue on version ~> 3.5.2

**Comment by darren-steven** [comment]
So here's a fun fact. Today, it is working perfectly. I walked away from my machine in frustration last night. Highly random behavior (every apply came up with different errors, or even worked). This morning (no changes to my config), everything is perfect.
I suspect the NerdGraph API might be struggling during the peak periods (maybe our account is throttled?)

**Comment by kidk** [comment]
We have a possible fix out with the latest provider v3.7.0. We changed our read mechanism to handle the API congestion. 

I'm going to close this for now. Please reply if you experience the error again and I will reopen.
